runners/jet/src/main/java/org/apache/beam/runners/jet/processors/WindowGroupP.java

```diff
@@ -0,0 +1,311 @@
+  }
+
+  @Override
+  public boolean complete() {
+    advanceWatermark(BoundedWindow.TIMESTAMP_MAX_VALUE.getMillis());
+    return flatMapper.tryProcess(EMPTY_BYTES);
+  }
+
+  private void advanceWatermark(long millis) {
+    this.latestWatermark = new Instant(millis);
+    this.keyManagers.values().forEach(m -> m.advanceWatermark(latestWatermark));
+  }
+
+  private static class InMemoryStateInternalsImpl extends InMemoryStateInternals {
+
+    InMemoryStateInternalsImpl(@Nullable Object key) {
+      super(key);
+    }
+
+    Instant earliestWatermarkHold() {
+      Instant minimum = null;
+      for (State storage : inMemoryState.values()) {
+        if (storage instanceof WatermarkHoldState) {
+          Instant hold = ((WatermarkHoldState) storage).read();
+          if (minimum == null || (hold != null && hold.isBefore(minimum))) {
+            minimum = hold;
+          }
+        }
+      }
+      return minimum;
+    }
+  }
+
+  /** Helper class that holds all the per key based data structures needed. */
+  private class KeyManager {
+
+    private final InMemoryTimerInternals timerInternals;
+    private final InMemoryStateInternalsImpl stateInternals;
+    private final ReduceFnRunner<K, V, Iterable<V>, BoundedWindow> reduceFnRunner;
+
+    KeyManager(K key, Instant currentWaterMark) {
+      this.timerInternals = new InMemoryTimerInternals();
+      this.stateInternals = new InMemoryStateInternalsImpl(key);
+      this.reduceFnRunner =
+          new ReduceFnRunner<>(
+              key,
+              windowingStrategy,
+              ExecutableTriggerStateMachine.create(
+                  TriggerStateMachines.stateMachineForTrigger(
+                      TriggerTranslation.toProto(windowingStrategy.getTrigger()))),
+              stateInternals,
+              timerInternals,
+              new OutputWindowedValue<KV<K, Iterable<V>>>() {
+                @Override
+                public void outputWindowedValue(
+                    KV<K, Iterable<V>> output,
+                    Instant timestamp,
+                    Collection<? extends BoundedWindow> windows,
+                    PaneInfo pane) {
+                  WindowedValue<KV<K, Iterable<V>>> windowedValue =
+                      WindowedValue.of(output, timestamp, windows, pane);
+                  if (ownerId.startsWith("3 ")) {
+                    new Throwable(
+                            WindowGroupP.class.getSimpleName()
+                                + " OUTPUT "
+                                + System.identityHashCode(WindowGroupP.this)
+                                + ", windowedValue = "
+                                + windowedValue)
+                        .printStackTrace();
+                  } // todo: remove
+                  // System.err.println(WindowGroupP.class.getSimpleName() + " OUTPUT ownerId = " +
+                  // ownerId + ", windowedValue = " + windowedValue); //useful for debugging
```

Comment:
Remove?

Reply:
Ok, will do so in the next commit (this PR).